### ISSUE TITLE
How does the mod settings code work?

### ISSUE THREAD

**Issue by aubergine10** [dymanoid/RealTime#125]
I'm creating a small mod (#117) which will require settings screen, and the implementation in RealTime looks impressive - particularly [RealTimeConfig.cs](https://github.com/dymanoid/RealTime/blob/master/src/RealTime/Config/RealTimeConfig.cs).

It looks like an architecture that many mods could benefit from. Are there plans to separate the config stuff in to separate repo or are there any docs on how to use it?

**Comment by dymanoid** [comment]
The settings UI code will be separated into its own infrastructure repository in scope of #84. 

There are no guides as such, but the public methods and attributes are documented.

The idea is: the config is defined by a simple container class that gets serialized to XML for storage. Special attributes define how to display various config values: which page, which group, what sort order, what UI element (check box, slider etc).

The code that actually creates a view is located in [the UI namespace](https://github.com/dymanoid/RealTime/tree/master/src/RealTime/UI).

**Comment by aubergine10** [comment]
Is there a way to configure it so that some options can be enabled/disabled based on other options?

**Comment by dymanoid** [comment]
Not yet, but it's certainly worth implementing.

**Comment by aubergine10** [comment]
Using attribs to specify default values might be worthwhile too?

```
using System.ComponentModel;

later...

        [ConfigItem(...)]
        [ConfigItemSlider(0f, 23.75f, 0.25f, SliderValueType.Time)]
        [DefaultValue(10f)]
        public float someHour { get; set; }

later...

 public void ResetToDefaults()
        {
            PropertyDescriptorCollection properties = TypeDescriptor.GetProperties(this);

            foreach (PropertyDescriptor property in properties)
            {
                AttributeCollection attributes = property.Attributes;

                DefaultValueAttribute defaultValue = (DefaultValueAttribute)attributes[typeof(DefaultValueAttribute)];

                if (defaultValue != null)
                {
                    this[property.Name] = defaultValue.Value;
                }
            }
        }

```

There's probably loads of errors in that, I still have much to learn about C#. But by using default val attribs it keeps all the interesting details about each config item together. It allows the reset to be more abstract (performance isn't really an issue for config objects) - it could be moved elsewhere to keep the config file super clean?

**Comment by aubergine10** [comment]
So, learnt a little more about C# and been working on my mod settings, these are the things I've found useful to add:

`Reset()`
I've reverted to using a method to reset values, as the junk in my comments above obviously won't work.
However, I'm setting default for each option inline, for example `public float IgniteHour { get; set; } = 10f;` as I find it useful to see what the default is with the definition. Maybe a custom attribute would be best approach?

`[ModOption(appMode)]` attribute
Clearly defines which options will end up in the .xml, and also which app mode they are applicable to (eg. game, editor, main menu, all, etc). I'm planning to use the appMode to determine whether an option gets persisted (e.g. only `Game` mode settings would be persisted to save games) and where an option gets displayed (eg. `None` mode, meaning "main menu", options shouldn't be shown on mod options screen when accessed from in-game pause menu).

`[Toggles(on[], off[])]` attribute
Indicates which other options will be enabled or disabled when a checkbox is ticked. Not happy with this one yet, as it's confusing to use and I'm rewriting it every few days with a different approach.

`UI components`
For the UI components, I've renamed their attributes to be more generic (not config specific) - while currently using them for config, I'm also expecting to use them for in-game UI also (eg. info panels) in future mods.

Also working on a slider variant that accepts a `typeof(someEnum)` or `string[]` whereby each step is one of the values in the list (a slider representation of a drop-down list; used when items denote a sliding scale).

`Log`
I'm splitting out the debug logging stuff in to separate file. It seems unnecessarily cumbersome having it merged in with the release logging implementation. I've also added a prefix feature to the log, ensuring mod name and version are always output at start of log messages.
